src/toolchain.rs

```diff
@@ -116,39 +116,88 @@ impl fmt::Display for DistToolchainDescription {
     }
 }
 
+#[inline]
+fn consume_back<T>(parts: &mut VecDeque<T>, number: usize) {
+    for _ in 0..number {
+        parts.pop_back();
+    }
+}
+
+/// Attempts to parse a date from the front of the parts list, returning the date and consuming the date parts if they are available
+fn extract_date(parts: &mut VecDeque<&str>) -> Option<Date> {
+    let len = parts.len();
+    if len < 3 {
+        return None;
+    }
+
+    let date_str = format!("{}-{}-{}", parts[len-3], parts[len-2], parts[len-1]);
+    match Date::parse(&date_str, DATE_FORMAT) {
+        Ok(d) => {
+            consume_back(parts, 3);
+            Some(d)
+        }
+        Err(_) => None,
+    }
+}
+
+fn extract_target(parts: &mut VecDeque<&str>) -> Option<TargetTriple> {
+    if parts.len() < 3 {
+        return TargetTriple::from_host().ok();
+    }
+
+    let len = parts.len();
+    let target_str = format!("{}-{}-{}", parts[len-3], parts[len-2], parts[len-1]);
+    match TargetTriple::new(&target_str) {
+        Ok(t) => {
+            consume_back(parts, 3);
+            Some(t)
+        },
+        Err(_) => {
+            if parts.len() < 4 {
+                return TargetTriple::from_host().ok();
+            }
+
+            let target_str = format!("{}-{}-{}-{}", parts[len-4], parts[len-3], parts[len-2], parts[len-1]);
+            match TargetTriple::new(&target_str) {
+                Ok(t) => {
+                    consume_back(parts, 4);
+                    Some(t)
+                },
+                Err(_) => TargetTriple::from_host().ok(),
+            }
+        }
+    }
+}
+
 impl FromStr for DistToolchainDescription {
     type Err = anyhow::Error;
     fn from_str(s: &str) -> Result<Self> {
```

Comment:
It would be good to add a comment about what formats of `s` are supported, since this is confusing

Reply:
To save time going back and forth, I updated the comment and tests to be more comprehensive: https://github.com/FuelLabs/sway/pull/5544

They're all passing so I'm happy to approve once those changes are merged in 😃